### ISSUE TITLE
DayTotalsPerEmployee not a resource error

### ISSUE THREAD

**Issue by metaclass-nl** [metaclass-nl/tutorial-api-platform#2]
Chapter 9 api 
With API Platform core 2.7 Hours operation get_day_report results in 
400 Error: Bad Request
`{
  "@context": "/contexts/Error",
  "@type": "hydra:Error",
  "hydra:title": "An error occurred",
  "hydra:description": "No resource class found for object of type \"App\\Model\\DayTotalsPerEmployee\".",`
(..)

[walkback](https://github.com/metaclass-nl/tutorial-api-platform/files/9792380/tutorial.issue.2.walkback.txt)

[This Api Platform core issue](https://github.com/api-platform/core/issues/5042) describes such a problem but for version 3.0. This is 2.7 with (metadata_backward_compatibility_layer: true), according to semantic versioning this should not happen. Furthermore, as [Oipnet](https://github.com/Oipnet) points out, the use of an output DTO is still present in the docs. However, it does not specify anything about the DTO class. The issue is closed with the argument that there is an other way to reach the same result. But this is not really the case for the report of day totals per employee: on a resource DayTotalsPerEmployee the standard orm filters can not filter the required hours with DayTotalsPerEmployee as resource class. Maybe passing Hours as resource class from the custom CollectionDataPovider  with the name of the normal collection operation get would work but this would be confusing because the filters of Hours would then be used for an operation of DayTotalsPerEmployee.

**Comment by metaclass-nl** [comment]
The problem is caused by the replacement of 
```php
// \ApiPlatform\Core\Serializer\AbstractItemNormalizer (from 2.6)
    public function normalize($object, $format = null, array $context = [])
    {
        if (!($isTransformed = isset($context[self::IS_TRANSFORMED_TO_SAME_CLASS])) && $outputClass = $this->getOutputClass($this->getObjectClass($object), $context)) {
            if (!$this->serializer instanceof NormalizerInterface) {
                throw new LogicException('Cannot normalize the output because the injected serializer is not a normalizer');
            }

            if ($object !== $transformed = $this->transformOutput($object, $context, $outputClass)) {
                $context['api_normalize'] = true;
                $context['api_resource'] = $object;
                unset($context['output'], $context['resource_class']);
            } else {
                $context[self::IS_TRANSFORMED_TO_SAME_CLASS] = true;
            }

            return $this->serializer->normalize($transformed, $format, $context);
        }
```
by:
```php
// \ApiPlatform\Serializer\AbstractItemNormalizer (from 2.7)
    public function normalize($object, $format = null, array $context = [])
    {
        $resourceClass = $this->getObjectClass($object);
        if (!($isTransformed = isset($context[self::IS_TRANSFORMED_TO_SAME_CLASS])) && $outputClass = $this->getOutputClass($resourceClass, $context)) {
            if (!$this->serializer instanceof NormalizerInterface) {
                throw new LogicException('Cannot normalize the output because the injected serializer is not a normalizer');
            }

            // Data transformers are deprecated, this is removed from 3.0
            if ($dataTransformer = $this->getDataTransformer($object, $outputClass, $context)) {
                $transformed = $dataTransformer->transform($object, $outputClass, $context);

                if ($object === $transformed) {
                    $context[self::IS_TRANSFORMED_TO_SAME_CLASS] = true;
                } else {
                    $context['api_normalize'] = true;
                    $context['api_resource'] = $object;
                    unset($context['output'], $context['resource_class']);
                }

                return $this->serializer->normalize($transformed, $format, $context);
            }
```
The old code looked for $outputClass to exest and IS_TRANSFORMED_TO_SAME_CLASS not to exist. If so it tried data trasforming. Without a data transformer the result (in our case supplied by the data provider) was allways normalized by the serializer. The new code only uses the serializer if a data transformer is actually to be applied.

**Comment by metaclass-nl** [comment]
Possible workaround: Make DayTotalsPerEmployee a resource:

```php 
// api/src/Model/DayTotalsPerEmployee.php
/**
 * Totals per day per Employee
 *
 * @ApiResource(
 *     itemOperations={
 *          "get"={
 *              "controller"=NotFoundAction::class,
                "read"= false,
                "output"= false
 *          }
 *      },
 *     collectionOperations={}
 * )
 */
class DayTotalsPerEmployee
(..)
```
After bin/console chache:clear this changes the output from:
```json
{
      "@type": "DayTotalsPerEmployee",
      "@id": "_:1005",
      "employee": {
        "@id": "/employees/19",
        "@type": "Employee",
        "label": "Eden, Nicky"
      },
      "from": "2019-09-18T00:00:00+00:00",
      "id": "19_2019-09-18 00:00:00",
      "label": "Eden, Nicky 2019-09-18",
      "to": "2019-09-19T00:00:00+00:00",
      "count": 2,
      "total": 8,
      "onInvoice": 8,
      "fractionBilled": 1
    }
```
to:
```json
{
  "@context": "/contexts/Hours",
  "@id": "/hours",
  "@type": "hydra:Collection",
  "hydra:member": [
    {
      "@id": "/day_totals_per_employees/19_2019-09-10%2000:00:00",
      "@type": "DayTotalsPerEmployee",
      "employee": {
        "@id": "/employees/19",
        "@type": "Employee",
        "label": "Eden, Nicky"
      },
      "from": "2019-09-10T00:00:00+00:00",
      "id": "19_2019-09-10 00:00:00",
      "label": "Eden, Nicky 2019-09-10",
      "to": "2019-09-11T00:00:00+00:00",
      "count": 1,
      "total": 7,
      "onInvoice": 7,
      "fractionBilled": 1
    },
```
However, this adds a resource to the api docs that is not really a resource and does not have a working operation 'get'.

**Comment by metaclass-nl** [comment]
For the time being to be solved by limiting chapters 9 and 10 to 2.6 in composer.json.

**Comment by metaclass-nl** [comment]
The problem has diappeared in 3.0. In 2.7 with     metadata_backward_compatibility_layer: false the error also goes away but the @type in the output is incorrect:
```json
   {
      "@type": "Hours",
      "@id": "/.well-known/genid/58cc41e59e1b7524cf57",
      "employee": {
        "@id": "/employees/19",
        "@type": "Employee",
        "label": "Eden, Nicky"
      },
      "from": "2019-09-10T00:00:00+00:00",
      "id": "19_2019-09-10 00:00:00",
      "label": "Eden, Nicky 2019-09-10",
      "to": "2019-09-11T00:00:00+00:00",
      "count": 1,
      "total": 7,
      "onInvoice": 7,
      "fractionBilled": 1
    },
```
while in 3.0 it is:
```json
    {
      "@type": "DayTotalsPerEmployee",
      "@id": "/.well-known/genid/3eb4f408887cca993627",
      "employee": {
        "@id": "/employees/19",
        "@type": "Employee",
        "label": "Eden, Nicky"
      },
      "from": "2019-09-10T00:00:00+00:00",
      "id": "19_2019-09-10 00:00:00",
      "label": "Eden, Nicky 2019-09-10",
      "to": "2019-09-11T00:00:00+00:00",
      "count": 1,
      "total": 7,
      "onInvoice": 7,
      "fractionBilled": 1
    },
```
A branch chapter10-api-3.0 has been pushed with code adapted to 3.0. However, it is still a lot of work to adapt all the other chapters including the readme's so for the time being the limiting to 2.6 will remain in place.
